# # Çoklu Doğrusal Regresyon ile Satış Tahmini 👩‍💻

**Satış Tahmini Modeli Geliştirme**

Bu çalışma, çeşitli reklam kanallarında yapılan harcamalara dayalı olarak satış tahmini yapmayı amaçlamaktadır. Modelimiz, TV, radyo ve gazete reklam harcamalarına ilişkin verileri kullanarak bu reklam harcamalarının satış üzerindeki etkisini tahmin edecektir.

Değişkenler:

*  TV: TV reklam harcamaları
* Radio: Radyo reklam harcamaları
* Newspaper: Gazete reklam harcamaları
* Sales: Satış miktarı


Bu bağlamda, çoklu doğrusal regresyon kullanarak satış tahmini işlemleri gerçekleştirilecektir. Modelin başarısı, tahmin sonuçlarının gerçek satış verileri ile karşılaştırılması yoluyla değerlendirilecektir.


**Gerekli kütüphaneleri içe aktarma**
Her zamanki gibi gerekli kütüphaneleri içe aktarmakla başlayacağız.

📌 "import" ve "from" anahtar sözcüklerini kullanın.

In [4]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option('display.float_format', lambda x: '%.2f' %x)

from sklearn.linear_model import LinearRegression
from sklearn.metrics import  mean_squared_error , mean_absolute_error
from  sklearn.model_selection import  train_test_split , cross_val_score

**Verileri Setini Yükleme ve Okuma**

.csv dosyasını okuyalım.

 Pandas kütüphanesinin read_csv() fonksiyonunu kullan


In [11]:
df = pd.read_csv("advertising.csv")

X = df.drop('sales' , axis=1)
y= df[["sales"]]

**Modelin Kurulması**

In [17]:
X_train , X_test , y_train , y_test = train_test_split(X , y, test_size = 0.20 , random_state= 1)

X_test.shape
y_test.shape

(40, 1)

In [20]:
reg_model = LinearRegression().fit(X_train, y_train)

#sabit (b- bias)
b = reg_model.intercept_
print("bias: " ,b )

#coefficients (w - weights)

w = reg_model.coef_
print("coefficients: " , w)


bias:  [2.90794702]
coefficients:  [[0.0468431  0.17854434 0.00258619]]


**Çoklu Regresyon Taahmin İşlemleri** :

Aşağıdak gözlem değerlerine göre satışın beklenen değeri nedir?
* TV : 30
* Radio : 10
* Newspaper : 40

Sabitler
* b : 2.90794702
* w : 0.0468431,   0.17854434 , 0.00258619



In [26]:

#sales = 2.90 + TV * 0.04 + radio * 0,17 + newspaper * 0,002

sales = 2.90794702 + 30 * 0.0468431 + 10 * 0.17854434 + 40 * 0.00258619

print(sales)



6.20213102


In [24]:
yeni_veri = [[30], [10] , [40]]
yeni_veri = pd.DataFrame(yeni_veri).T

reg_model.predict(yeni_veri)

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([[6.202131]])

**Çoklu Doğrusal Regresyon Tahmin Başarısı 🙂**

In [31]:

#Train RMSE
y_pred = reg_model.predict(X_train)
rmse = np.sqrt(mean_squared_error(y_train , y_pred))
print(f'RMSE: {rmse}')

#Train RKare
rk = reg_model.score(X_train , y_train)
print(f'R-Kare: {rk}')

#Test RMSE
y_pred = reg_model.predict(X_test)
rmse_t = np.sqrt(mean_squared_error(y_test , y_pred))
print(f'RMSE_T: {rmse_t}')

#Test R Kare
rk_t = reg_model.score(X_test , y_test)
print(f'R-Kare_T: {rk_t}')




RMSE: 1.736902590147092
R-Kare: 0.8959372632325174
RMSE_T: 1.4113417558581587
R-Kare_T: 0.8927605914615384



Doğrusal Regresyonda elde ettiğimizde başarı değerleri:  
* RMSE: 3.2423221486546887
* R-Kare: 0.611875050850071

📌⭐ "Doğrusal regresyonda elde ettiğimiz başarı değerlerini kıyasladığımızda, *RMSE: 3.24* ve *R-Kare: 0.61* sonuçları ortaya çıkarken, çoklu doğrusal regresyonda bu değerler *RMSE: 1.74* ve *R-Kare: 0.90* olarak iyileşmiştir. Bu da, çoklu doğrusal regresyonun modelin performansını önemli ölçüde artırdığını göstermektedir."

Üstelik, test hatasının 1.41, eğitim hatasından 1.73 daha küçük çıkması oldukça olumlu bir senaryodur. Test seti için R-Kare değerinin 0.89 olması, bağımsız değişkenlerin, bağımlı değişkeni açıklama oranının yaklaşık %90 olduğunu göstermektedir; bu da oldukça yüksek bir yetenek seviyesine işaret eder.


In [49]:
#10 Katlı CV RMSE
rmse = np.mean(np.sqrt(-cross_val_score(reg_model,
                                        X,
                                        y,
                                        cv=10,
                                        scoring="neg_mean_squared_error")))
print("10 Katlı CV RMSE:", rmse)


#5 Katlı CV RMSE
rmse = np.mean(np.sqrt(-cross_val_score(reg_model,
                                        X,
                                        y,
                                        cv=5,
                                        scoring="neg_mean_squared_error")))
print("5 Katlı CV RMSE:", rmse)





10 Katlı CV RMSE: 1.6913531708051797
5 Katlı CV RMSE: 1.7175247278732086
